unix/unix_cstruct.ml

```diff
@@ -17,3 +17,60 @@
 let of_fd fd =
   let buffer = Bigarray.(array1_of_genarray (Unix.map_file fd char c_layout false [|-1|])) in
   Cstruct.of_bigarray buffer
+
+type buffer = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
+
+(* Returns 0 if there is no writev *)
+external stub_iov_max: unit -> int = "stub_cstruct_iov_max"
+
+external stub_write: Unix.file_descr -> (buffer * int * int) -> int = "stub_cstruct_write"
+
+external stub_writev: Unix.file_descr -> (buffer * int * int) list -> int = "stub_cstruct_writev"
+
+let iov_max = stub_iov_max ()
+
+(* return the first n fragments, suitable for writev *)
+let rec first n rev_acc = function
+| [] -> List.rev rev_acc
+| _ when n = 0 -> List.rev rev_acc
+| x :: xs -> first (n - 1) (x :: rev_acc) xs
+
+(* shift a list of fragments along by n bytes *)
+let rec shift t x =
+  if x = 0 then t else match t with
+  | [] -> invalid_arg "foo"
+  | y :: ys ->
+    let y' = Cstruct.length y in
+    if y' > x
+    then Cstruct.shift y x :: ys
+    else shift ys (x - y')
+
+let rec write fd buf =
+  if Cstruct.length buf > 0 then begin
+    let n = stub_write fd (buf.Cstruct.buffer, buf.Cstruct.off, buf.Cstruct.len) in
+    write fd @@ Cstruct.shift buf n
+  end
+
+let writev fd bufs =
+  let rec use_writev = function
+    | [] -> ()
+    | remaining ->
+      (* write at most iov_max at a time *)
+      let to_send = first iov_max [] remaining in
+      let n = stub_writev fd (List.map (fun x -> x.Cstruct.buffer, x.Cstruct.off, x.Cstruct.len) to_send) in
```

Comment:
do you expand to a tuple here because the Cstruct.t record type might change? Given this is in the same repository, it's probably safe enough to just pass the Cstruct record in directly to the FFI and save some allocations

Reply:
Yes that was it -- I was using these stubs in another repo. I'll drop the tuple since we're here.